HackyAI - New AvoidCapturing, Multitarget SupportPower support, and LimitUnitBuildingByActors

StormWingDelta opened the pull request:
I'll admit I got a tad carried away on this one so I'll try to summarize as best I can.

AvoidCapturingActorTypes: this now allows the programmer in the yaml to name what actor types should be avoided when capturing things, this can be used alongside the CapturableActorTypes to help narrow down the list of targets further although using both at the same time is possible it's overkill.

Multitargetting Support Powers: This has always been an issue for the AI since it assumes the support power only has one target and doesn't specify other targets.  This was resolved through the use of Named Considerations where you name each of the targets you are making considerations for.  This allows the use of things like the Chronosphere that needs 2 targets to fully work instead of one.  I'm unable to add 3+ target support at this time due to not having any examples of support powers that use 3 more more targets.  Under SupportPower I added a NumberOfTargetsOrSteps to allow both the AI and support power to have something for later that could open the door to N Target support powers.  Currently it defaults to 1 for all support powers and has to be changed in the YAML of the actor using a support power to another number.

LimitBuildingByActors and LimitUnitByActors: Yes I know they could both use better names and are basically the same code if anyone has ideas let me know.  Both of these allow the AI through the yaml to limit the amount of a given unit or structure it makes based on the amount of other actors it has made already.  This allows for say limiting the production of MCVs based on the amount of Construction Yards for example.  This technically can help anything that uses the code for transforming or deploying if it takes on a completely differently named form thus not existing as what it was before but it can be used for other things as well to act as a pseudo build order for starting and stopping production of things.

I figured I'd get these in before I went completely overboard so I could make sure it didn't conflict with anything and maybe improve on it further.  I have an AI named Impulse that demonstrates how to use the new things the AI has use of if anyone wants it.  That said all the code should function without any of the additions I made needing to be there.

reaperrr committed:
Make Explodes conditional

reaperrr committed:
Add elite explosion to TS Sonic & Stealth Tanks

Matching original TS.

reaperrr committed:
Tweak TS sprite walker animations

- replace WithInfantryBody with WithFacingSpriteBody + WithMoveAnimation on Titan
- rename all mech 'run' anims to 'walk'
- tweak walk anim rates to better match movement speed/original
- tweak Wolverine attack anim to play a little faster

reaperrr committed:
Fix TS Titan barrel, armament offsets and disable projectile image

MustaphaTR committed:
Reimport 8PLAY3 and 5

reaperrr committed:
Change Explodes.Weapon default to null

FieldLoader.Require makes the default pointless, so it makes more sense to change it to null.

jrb0001 committed:
Make SpawnActorOnDeath conditional

StormWingDelta committed:
Merge remote-tracking branch 'refs/remotes/OpenRA/bleed' into bleed

StormWingDelta committed:
Merge remote-tracking branch 'refs/remotes/OpenRA/bleed' into bleed

StormWingDelta committed:
Merge remote-tracking branch 'refs/remotes/OpenRA/bleed' into bleed

StormWingDelta committed:
Merge remote-tracking branch 'refs/remotes/OpenRA/bleed' into AI_Update1

StormWingDelta committed:
Merge remote-tracking branch 'refs/remotes/OpenRA/bleed' into AI_Update1

StormWingDelta committed:
Merge remote-tracking branch 'refs/remotes/OpenRA/bleed' into bleed

StormWingDelta committed:
Merge remote-tracking branch 'refs/remotes/origin/bleed' into AI_Update1

StormWingDelta committed:
Added Support for Multitarget support powers for the AI

This allows the AI to be able to target with support powers that require
more than one target before they fire.

StormWingDelta committed:
Added as support for limiting production based on the amounts of other actors

This allows the AI to limit the production of units and structures based
on the amounts other units and structures on the map.

StormWingDelta committed:
Added AvoidCapturing, Multitarget SupportPowers, and ProductionLimits based on other actors

HackyAI can now Limit production based on other actors, tell engineers
what not to capture and make use of supportpowers that target multiple
locations before firing.

StormWingDelta committed:
Added NumberOfTargetsOrSteps to the Chronosphere power

Just added this to it so the AI would know what the heck to do when it
tried to use it.

StormWingDelta committed:
Added Copyright & Licence info

I forgot it

StormWingDelta committed:
Added Copyright and Licence Info

I forgot it

StormWingDelta committed:
Fixed StyleCop Complaints 

Ya it seems mine didn't catch some of these

StormWingDelta committed:
Fixed StyleCop Issues

>.>

StormWingDelta committed:
Fixed StyleCop Errors

Hopefully I got them all

StormWingDelta commented:
Unable to find that last StyleCop Error or fix it because Github won't let me commit what is basically no change.  Also I've got no clue what AppVeyor is complaining about.  What exactly is it wanting me to do?

StormWingDelta commented:
1) The first one I could swear I fixed with the tab stop. Also github doesn't seem to be able to tell it's there either and was preventing the fix to that line.  2) I'll try the two commands you mentioned and see what happens.  3) I guess I'll see about uploading all the changed files at once and see if that fixes it, but I have no idea how to make commit comments per file(s) without sending them up at different times.  4) Had no idea of the make check thing. >.>

LipkeGu commented:
When you use Vidual studio 20** there is a StyleCop extension for it ^^

StormWingDelta commented:
I have the StyleCop Extension it didn't pick up some of those errors the GitHub things you guys are using does. >.<

abcdefg30 commented on OpenRA.Mods.Common/AI/LimitUnitByActors.cs:
```diff
@@ -1,4 +1,15 @@
-﻿using System;
+#region Copyright & License Information
+/*
+ * Copyright 2007-2017 The OpenRA Developers (see AUTHORS)
```
2018

StormWingDelta replied:
OK I'll make sure these are and stay 2018 over the main files. :)

LipkeGu replied:
please do not remove this ```using System``` it is important for travis

StormWingDelta commented:
Also before I get yelled at for screwing up the loading of an AI trait are my Load Methods for both those new classes even remotely correct?

StormWingDelta commented:
Unable to remove this: OpenRA.Mods.Common/AI/HackyAI.cs:L566: [SP2000] Invalid spacing at the end of the line.

I've tried several times to fix this error and I'm unable to do so no matter what I try.

StormWingDelta committed:
Added a number of features explained in the PR and fixed the CR

I think I've fixed all the problems that were detected if this still
breaks I'll force upload the files through the cmd and check than.

StormWingDelta commented:
No idea how to solve the  AppVeyor Error but everything else is fixed and ready.  Anyone got ideas on how to fix this annoying error?

LipkeGu commented:
![neue bitmap](https://user-images.githubusercontent.com/2057932/34994853-17101f1e-fad5-11e7-8701-e701e18424ce.png)

Did you commit have commit ALL your changes ?? i cant find the class =(

LipkeGu commented on OpenRA.Mods.Common/AI/SupportPowerDecision.cs:
```diff
@@ -111,13 +141,57 @@ public int GetAttractiveness(IEnumerable<FrozenActor> frozenActors, Player fired
 
+			/// CtrlF - this was done to help with multistep support powers since each step would need its own target considerations
+			[Desc("Used to name the Consideration in the case of multistep support powers, Leave null or blank for code to ignore.")]
+			public readonly string ConsiderationName = null;
```
Use ```String.Empty``` for this cases

abcdefg30 replied:
`null` should be fine as default. We use it in a lot of other places as well.

GraionDilach replied:
Yeah, we're using IsNullOrEmpty where the string's used instead.

StormWingDelta replied:
Ya kinda spaced on what to use there so I'll make sure that gets fixed.

LipkeGu replied:
use null

LipkeGu commented on OpenRA.Mods.Common/Traits/SupportPowers/SupportPower.cs:
```diff
@@ -73,7 +73,21 @@ public abstract class SupportPowerInfo : PausableConditionalTraitInfo
 
-		public SupportPowerInfo() { OrderName = GetType().Name + "Order"; }
+		// Added NumberOfTargetsOrSteps for AI Multiple SupportPower targeting and future N Target Supports
+		// ~~StormWing0 made this for AI usage mainly but hopefully at some point it can be used for more than that.
```
Its not important who added it..... remove this line please.

StormWingDelta replied:
Almost couldn't tell what line you were talking about there for a moment.  Also if there's a way to get around needing the NumberOfTargetsOrSteps line itself please do hit me over the head with ideas because that's the only thing I could think of at the time that'd get the effect I was looking for while still leaving the door open for other things down the road.

LipkeGu reviewed (changes requested):
I checked out your branch and found your compilation error -> Please add the two missing references to the classes or Types in your IDE by using "Add existing..." (this will fix the compilation.) 

![addexsisting](https://user-images.githubusercontent.com/2057932/34999499-1bdbfc54-fae2-11e7-8afa-e5e824e55a08.png)

and try to compile it locally by rebuilding the entire Solution or run via command line:


```
make clean
make dependencies
make all
make check
```

Phrohdoh commented on OpenRA.Mods.Common/AI/LimitBuildingByActors.cs:
```diff
@@ -0,0 +1,281 @@
+			return ret;
+		}
+
+		// Checks if the Unit has any limiters telling the AI when to start making it
```
What @LipkeGu is referring to here is colloquially called [xmldoc](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/xmldoc/how-to-use-the-xml-documentation-features) which many .NET tools use to aide developers (there are even tools which generate HTML documentation from these comments).

StormWingDelta replied:
On it, want me to add summary comments anywhere else (Besides my edit spots) while I'm at it in the files I've edited?

abcdefg30 commented on OpenRA.Mods.Common/AI/LimitBuildingByActors.cs:
```diff
@@ -0,0 +1,281 @@
+		{
+			if (LimitedBuildings != null)
+			{
+				var limitergroup = LimitedBuildings.Where(ul => ul.LimitedBuilding != null && ul.LimitedBuilding.ToLower().Equals(unitName.ToLower()));
```
What about using `StringComparison.OrdinalIgnoreCase` on `Equals` instead?

pchote replied:
Why does this need to use `.Equals` at all?  This isn't java, we have a working `==` operator :smile:

Phrohdoh replied:
You cannot perform allocation-free case-insensitive comparisons with `==`.

pchote replied:
We don't want to be doing case-insensitive comparisons at all...

StormWingDelta replied:
Technically for me it's an old habit from my days using Java and not having the darnedest clue why my code wasn't working.  If you like it the other way I can set it that way. :)

LipkeGu replied:
strings are allocated in C# just as like in C++, ages where you have to allocated space for a null terminated char array as in C are gone :) simply use ```==```

GraionDilach commented:
This definitely needs a rebase as well to remove the dummy merge commits.

GraionDilach commented on OpenRA.Mods.Common/AI/HackyAI.cs:
```diff
@@ -248,6 +264,40 @@ static object LoadDecisions(MiniYaml yaml)
 		}
 
+		// fair warning I wasn't sure how to set this up so if someone can get it setup right show me so it doesn't happen again. :) ~~StormWing0
+		static object LoadUnitAdditionalUnitLimits(MiniYaml yaml)
```
These look really weird but I'm not certain at this point how these are supposed to look in the YAML. Could you please write an example setup with limited MCVs based on conyards and limiting, say, the helis on each other?

StormWingDelta replied:
I've got an example on my end, I'll upload the yaml AI I was using to test the code later if wanted but here's the sample you wanted: 

LimitUnitByActors:
	UnitLimit:
		LimitedUnit: mcv
		#CheckAllStopLimiters: true
		StopProductionLimiters:
			fact: 3

It's the same idea for the LimitBuildingByActors as well. :)

StormWingDelta replied:
Is there anyone that has ideas on how to write both these new loading methods better?  I was trying to write it so the new code didn't have to be used in the yaml to load without error but didn't have any examples I could find of loading non-lists from the yaml.

GraionDilach commented on OpenRA.Mods.Common/AI/HackyAI.cs:
```diff
@@ -434,20 +484,41 @@ public void QueueOrder(Order order)
 
-			var unit = buildableThings.Random(Random);
+			// ~~StormWing0 Added new checks for if a unit can be built or trained and moved some older checks into it
+			var availibleUnits = buildableThings.Where(actorUnit =>
```
avalaibleUnits

StormWingDelta replied:
oops didn't see that. >.<

StormWingDelta replied:
Oh I remember what I was trying to do. The reason I went that route was I wanted the AI to check what it could build for units first before trying to build the unit since it kept locking up the AI attempting to build the same unit over and over again.  So I figured I'd make it so it'd only pick from what it could build instead of picking from everything than checking if it could make it.

I think I can narrow down the lines of code to cut out a ton of that junk so it doesn't look so odd.

GraionDilach commented:
I would honestly split out the three logic extensions into three separate PRs.

LipkeGu commented:
I would suggest to splitt this in 3 commits (the last is the final implementation and the two ohters the introduction.

StormWingDelta commented:
Oh sorry about being gone so long I'll try and get all the changes in as soon as able. Also if wanted I can upload my additional AI I was using to test the new code if anyone wants it. :)

StormWingDelta commented on OpenRA.Mods.Common/AI/BaseBuilder.cs:
```diff
@@ -297,13 +328,18 @@ ActorInfo ChooseBuildingToBuild(ProductionQueue queue)
 
+				// Added new checks for if a building can be built and moved some older checks into it
+				// CTRLF - Check if the building has reached its limits
+				if (!BuildingIsBuildable(name))
```
Ya it took a couple of tries to get it working when figuring out how that area of the code ran.  As for the AppVayor Error I don't get any errors on my end when compiling about the new class files so it might have to do with how I upload.  Either that or the error is going silent on my end one of the other.

StormWingDelta commented on OpenRA.Mods.Common/AI/LimitBuildingByActors.cs:
```diff
@@ -0,0 +1,281 @@
+		}
+
+		// Checks if the Unit has any Limiters telling the AI to stop making it
+		public bool BuildingHasStopLimiters(string unitName)
```
Ya there's a reason for it, while it looks like it's a duplicate the use of the start and stop limiters check it opposite.  Start returns false as long as it is below the limit and Stop returns false as long as it is at or above the limit.  I'll see if I can make it not so duplicate appearing once able. I did have the code all in one method for the check but it didn't seem to play nice for whatever reason so I split it into copies for both start and stop checks.  If I can make them back into one method for both checks I will unless someone else beats me to it.

LipkeGu replied:
When it isnt able then let it as is...

StormWingDelta replied:
One of my code passes I got around it by using an extra parameter to change between the two targets. :) I'll see if I can get that setup again.

StormWingDelta commented on OpenRA.Mods.Common/AI/SupportPowerDecision.cs:
```diff
@@ -111,13 +141,57 @@ public int GetAttractiveness(IEnumerable<FrozenActor> frozenActors, Player fired
+		}
+
+		/// <summary>Evaluates the attractiveness of a group of actors according to all considerations of the selected name</summary>
+		/// CtrlF
```
Ya I'll remove those CtrlF things they were mostly for me when trying to get around fast to areas I was working on.

LipkeGu replied:
ok, you mean ```search for``` your IDE has the ability for it... called bookmarks see here how you can use it... https://www.youtube.com/watch?v=JhxC-K-Eehg

GraionDilach commented:
You'd probably end up better if you push this revision with your example AI into another branch then clean up this one to only have a single feature. That way we can still get a full insight on your long-term plan but still be able to properly review your code, because this is a huge PR as-is with many picks need to be worked on due to your lack of experience with our coding standards and policies.

pchote commented:
Please also note the text in the text box when you file a PR.  That contains a lot of info about how you can make sure your branch is ready to review before others have to spend any time on it.

StormWingDelta commented:
I'll shoot for the simplest 2 first that aren't causing the main error than since they both use existing files and shoot up the one that adds files last since I'm not sure what's up with it throwing the error despite not on my end.
So it should go like this:

1. AI Avoid Capturing Actors code
2. AI Able to use multi-targeting Support Powers (Has everything already but want to add something to let it account for other support power target locations)
3. AI Limiting Production of Actors by number of other actors (Has everything except maybe narrowing down by faction)
4. AI Able to Limit the max of an actor's production based on a ratio with another actor (No idea at this point how this one is going to work)
5. AI able to use any Actor Ability Trait when told to do so and if needed how to do so (Really no idea on this one since my last attempt failed due to no way to grab a trait/traitinfo from a string that I know of)

That should work since it goes from easiest to hardest.  After that I have no idea what to do in the case of the AI.  My main goal is to make the AI able to do addition things it can't currently but the player is able to do.  Like knowing what to capture and what to avoid, able to handle support powers that hit more than one target before firing, able to know how to use actor abilities, etc.  Also making it seem like it can change strategy based on what's going on for it and things around it.

MustaphaTR commented:
> AI able to use any Actor Ability Trait when told to do so and if needed how to do so (Really no idea on this one since my last attempt failed due to no way to grab a trait/traitinfo from a string that I know of)

You can use `world.Map.Rules.Actors["string"]` to get an `ActorInfo` as seen [here](https://github.com/OpenRA/OpenRA/blob/8a6091884127ea2c6706662332ee70097c81885e/OpenRA.Mods.Common/Orders/PlaceBuildingOrderGenerator.cs#L58).

StormWingDelta commented:
The issue is the next line with the Generic since you have to put in something known instead of something you don't know until someone names it.

StormWingDelta commented on OpenRA.Mods.Common/AI/BaseBuilder.cs:
```diff
@@ -297,13 +328,18 @@ ActorInfo ChooseBuildingToBuild(ProductionQueue queue)
 
-				if (ai.Info.BuildingLimits.ContainsKey(name) && ai.Info.BuildingLimits[name] <= count)
-					continue;
+				// if (ai.Info.BuildingLimits.ContainsKey(name) && ai.Info.BuildingLimits[name] < count)
```
On it, I think I got most of the code I commented out, just forgot to remove bits no longer needed.

StormWingDelta commented:
Uploading in a more sensible manner from a second branch feel free to close this PR.

abcdefg30 commented:
Superseded by #14756 (and follow-ups).

StormWingDelta commented:
TY hopefully I get things right this time. :)
